doc/man3/OSSL_PARAM_BLD.pod

```diff
@@ -32,6 +32,8 @@ OSSL_PARAM_BLD_push_octet_string, OSSL_PARAM_BLD_push_octet_ptr
                             const BIGNUM *bn);
  int OSSL_PARAM_BLD_push_BN_pad(OSSL_PARAM_BLD *bld, const char *key,
                                 const BIGNUM *bn, size_t sz);
+ int OSSL_PARAM_BLD_push_signed_BN(OSSL_PARAM_BLD *bld, const char *key,
+                                   const BIGNUM *bn);
```

Comment:
Discussion point:

I struggled with deciding on the "proper" API for BIGNUMs.  The options I came up with were:

1.  A form that would be more like the other integer pushing functions.  Unfortunately, since `OSSL_PARAM_BLD_push_BN()` already exists, it would mean adding a `_ex` variant:

    ``` C
    /* Pushes a BN that gets the data type OSSL_PARAM_INTEGER */
    int OSSL_PARAM_BLD_push_BN_ex(OSSL_PARAM_BLD *bld, const char *key,
                                  const BIGNUM *bn);
    /* Pushes a nonnegative BN that gets the data type OSSL_PARAM_UNSIGNED_INTEGER */
    int OSSL_PARAM_BLD_push_uBN(OSSL_PARAM_BLD *bld, const char *key,
                                const BIGNUM *bn);
    ```

2.  Just add a new function that handles BNs as signed integers.  This is contrary in pattern to the other TYPE pusher, but doesn't add an `_ex` function:

    ``` C
    /* Pushes a BN that gets the data type OSSL_PARAM_INTEGER */
    int OSSL_PARAM_BLD_push_signed_BN(OSSL_PARAM_BLD *bld, const char *key,
                                      const BIGNUM *bn);
    ```

Reply:
What about having OSSL_PARAM_BLD_push_BN() decide which type to use based on the sign of the BN?  If it is negative, push an integer, if not push an unsigned.

The other helper functions should make the dichotomy invisible.

Reply:
So what you're saying is that if the receiving end expects an OSSL_PARAM_INTEGER, they must always also expect an OSSL_PARAM_UNSIGNED_INTEGER?  How would a `_gettable` or `_settable` express that in its returned OSSL_PARAM descriptor array?

Also, that's conceptually confusing, as ℤ (which is essentially what OSSL_PARAM_INTEGER represents) does include positive numbers.  It's further confusing, because no other of the numeric OSSL_PARAM functions behave this way.  Why don't you propose that OSSL_PARAM_BLD_push_int32() should do this as well, i.e. make positive numbers OSSL_PARAM_UNSIGNED_INTEGER instead of OSSL_PARAM_INTEGER?

Reply:
I'm suggesting extending the internal type conversions to OSSL_PARAM_get_BN() and OSSL_PARAM_set_BN() when handling integer arguments.  I.e. they should accept either signed or unsigned and convert appropriately (or error if it isn't possible).  This is the same as all the other helpers -- you are free to set an int into a size_t param and it just works (or errors).

This would then flow down in into the builder etc where either type could be used.  Integer for negative, unsigned integer for positive then becomes workable and sensible.

As for the params definitions, this is user facing and is natural: one that can be negative is an integer, one that can't is an unsigned integer.  Perfect from the user's viewpoint -- they define what they want and our code makes it happen.

I don't consider this is in the least bit confusing.  Slightly more complicated to implement in places perhaps, but not confusing -- things that just work rarely are confusing.  That is what users really care about.  Nobody cares what's going on underneath.

Reply:
Please avoid allegations on what other people care about, there's no way you can know.

Assuming that everyone does use the OSSL_PARAM functions (which no one is *obligated to*, I'll remind you), we would have to extend other functions as well to accept both OSSL_PARAM_INTEGER and OSSL_PARAM_UNSIGNED_INTEGER, at least these these:

``` C
int OSSL_PARAM_get_int(const OSSL_PARAM *p, int *val);
int OSSL_PARAM_get_long(const OSSL_PARAM *p, long int *val);
int OSSL_PARAM_get_int32(const OSSL_PARAM *p, int32_t *val);
int OSSL_PARAM_get_int64(const OSSL_PARAM *p, int64_t *val);
```

Possibly also this, if `time_t` is signed:

``` C
int OSSL_PARAM_get_time_t(const OSSL_PARAM *p, time_t *val);
```

Reply:
I suggest you have a look at the code for those functions.  They **all** accept OSSL_PARAM_INTEGER, OSSL_PARAM_UNSIGNED_INTEGER and OSSL_PARAM_REAL and convert as appropriate (if the conversion is exact).  The BN functions should do the same but don't (yet).

Have a look at our code base and count the number of times the helper functions aren't used: I'm only aware of one test which does things directly (and I _hate_ that test, it's a mess).

I can make a good guess as to what users' want: things being easy.  What you are proposing isn't as easy to use as what I'm suggesting.

Reply:
You know, thanks for the reminder, I had forgotten all the type coercion work.

> I'm only aware of one test which does things directly (and I _hate_ that test, it's a mess).

You may hate it, but it's important, 'cause it's intentional that others should be able to handle OSSL_PARAM raw, for whatever reason they choose.

Reply:
I am with Pauli on this one. I do not understand why having a single (or two - padded and unpadded) function makes raw handling of OSSL_PARAM(s) harder.

Yeah if you create a param this way it can be either signed for negative numbers or unsigned for positive ones and the receiving side would have to be able to handle both types.

Maybe, as a compromise the OSSL_PARAM_BLD_push_BN should do this automatic type change but there should be also a OSSL_PARAM_BLD_push_signed_BN that would always use the OSSL_PARAM_INTEGER.

Yeah, when using OSSL_PARAM_BLD_push_BN() with negative BIGNUM with the receiver that does not accept the negative value, the error would be recognized a little bit later, but does that really matter?

Reply:
Awright.

Related question: would it be feasible to treat, say, int the same way (i.e. modify OSSL_PARAM_BLD_push_int())?  That is essentially equivalent!

Reply:
If "no", why not?

Reply:
No, because the push_int already has an existing semantics of creating an INTEGER param and not UNSIGNED INTEGER for positive numbers.

Where the existing semantics of push_BN to error out for negative bignums could be extended naturally. (Not speaking also about the thing that the erroring out was added just recently and was not there with 3.0.0 where it just produced an incorrect number.)

Reply:
So, how do you all propose this gets documented?
Whatever we do, push_BN becomes a bit exceptional compared to the other integer pushing functions. I find that unfortunate.

Reply:
The only documentation is for the push_BN function that says something along the lines of pushing an integer for negative values and an unsigned integer for the rest.

push_BN is already quite different from from the rest -- it pushes a variable sized integer.

In answer to an earlier question: look at the code for the various get/set OSSL_PARAM functions -- they already deal with integer, unsigned integer and reals just fine.  Make it simple and flexible and users will like it.

How many times are raw params used instead of the helpers?  The only instance I can think of is one unit test (one which I've grown to dislike BTW).

Reply:
Whatever.

I've made the changes, even though I strongly dislike it.  And documented.